Add: allow historical maps to be selected in settings

SlySven opened the pull request:
#### Brief overview of PR changes/additions
This adds a `QComboBox` selector under the load map button in the mapper tab on the "Profile preferences" dialogue to allow any map (Mudlet `.dat` binary, `.json` or `.xml`) file stored in the "normal place" to be chosen and loaded. Like the game save selection in the "Connect profile" dialogue an icon is pre-pended to the "autosave" one, in addition a (icon theme dependent) icon is pre-pended for `.json` and `.xml` files so they can be distinguished from the "normal" binary one. Like that game save list the entries are sorted into descending date and time order.

Also, renames:
* `(QComboBox*) dlgConnectProfiles::profile_history` ==> `comboBox_profileHistory`

For optimisation reasons the `QRegularExpression`s used to detect the details of normal game and map save files have been made into `static const`s so that they can be optimised and created just once, rather than every time a different profile in the "Connect profile" dialog is opened.

#### Motivation for adding to Mudlet
It is a bounty issue.

#### Other info (issues closed, discussion etc)
This should close #656!

This PR originally ALSO added a similar `QComboBox` to the second tab of the "Connect profile" dialogue - which I thought was a good idea but which was declared to be outside of the scope of the original issue. I will promulgate that option again in a future PR where we can discuss how good an idea it really is! :smiling_imp:

SlySven committed:
Add: allow historical maps to be selected on opening profile

This should close #656!

This adds another `QComboBox` selector under the game save selector in the
second (options) tab on the "Connect profile" dialogue to allow any map
(Mudlet `.dat` binary, `.json` or `.xml`) file stored in the "normal place"
to be chosen and loaded instead of the default "last" one. Like the game
save selection an icon is pre-pended to the "autosave" one, in addition a
(icon theme dependent) icon is pre-pended for `.json` and `.xml` files so
they can be distinguished from the "normal" binary one. Like the game save
list the entries are sorted into descending date and time order.

Also, renames:
* `(QComboBox*) dlgConnectProfiles::profile_history`
                                              ==> `comboBox_profileHistory`

For optimisation reasons the `QRegularExpression`s used to detect the
details of normal game and map save files have been made into
`static const`s so that they can be optimised and created just once, rather
than every time a different profile in the "Connect profile" dialog is
opened.

It became clear that some lines for odd files could be really long and
something to adjust the widths of some widgets in the "Connect profile"
dialog would be beneficial. I have added code that tries to do that, it
should ensure that there is enough space on the left to show three columns
of profile icons and enough on the right to show the full length of the
longest file selected in the two combo boxes for the game save and map
files for the currently selected profile.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (commented):
Amazing! Thanks, happy to see this.

Could you move this to the preferences - mapper tab through? That's what the issue desires too.

SlySven commented:
> Could you move this to the preferences - mapper tab through? That's what the issue desires too.

Can you clarify that - do you mean you want the **same** functionality implemented on the Mapper tab in the preferences - which will be **separate** from the button to "Load" a map as that can look ***anywhere in the filesystem***?

vadi2 commented:
That's right. It should be either possible to load an older version of the map from the dropdown or load an entirely different map file anywhere from the filesystem.

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

SlySven committed:
Revise: add historic map loading to profile preferences as well

This almost clones the code in `dlgConnectionProfiles.cpp`.

Also "cleans-up" the form/dialogue file for the class - convert
`QGridLayout`s to either `QVBoxLayout` or `QHBoxLayout` as appropriate and
put the remaining ones in item order. Fix up the tab order to accommodate
recently added widgets.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: prevent crash from trying to access a non-existent mapper with XML map

This seems to be possible if loading an XML map at profile startup.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: ensure that all map types loaded from preferences get audited

So that the `[ OK ]` message gets put on main console.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

Conflicts resolved in:
* src/ui/connection_profiles.ui

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: ensure map loading does not take place until Lua sub-system is started

The `TMap::audit()` function - which produces the `[ OK ]`  message for a
map loading operation makes use of an external Lua function, so it cannot
be used until the Lua sub-system has been set up. Doing the map loading
from a zero-timer at the end of the `Host` constructor is too early to allow
this to work properly - so it needs to be moved to something that can be
called after then.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revert: remove stuff from connect profile dialogue. 8-(

Not wanted by peers in this PR.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

vadi2 committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

vadi2 reviewed (commented):
![image](https://user-images.githubusercontent.com/110988/229731277-a81b31ae-7e63-44bb-b2f4-66fde6ec8022.png)

It looks a bit awkward right now - what about this simpler design:

```
Load another map file in: [choose map file] or select another version of the existing map [combobox]
```

@Mudlet/ui-review

Kebap commented:
The original post here is outdated. Which changes are intentional? 
The ui file diff lists lots changes unrelated to loading maps. 
I don't like the new horizontal lines added. Makes the tab even more cluttered. 

> what about this simpler design

How would you load the map selected in dropdown? 

Great idea to discuss and align on design ideas before producing actual code!

vadi2 commented:
> How would you load the map selected in dropdown?

By selecting it from the dropdown.

SlySven commented:
>  `Load another map file in: [choose map file] or select another version of the existing map [combobox]`

How would that be fitted into the space? :confused: That `QComboBox` needs to be wider than one column to accommodate both long (localised) Date and time formats (for `en_US` locales: ` dddd, MMMM d, yyyy h:mm:ss AP`;  the alternative short form is, for, `en_US` locales `M/d/yy h:mm AP` which is less easy to put into mind as, oh that last save from Wednesday night) AND other odd files with arbitrary names that might be put in that location:

![image](https://user-images.githubusercontent.com/6163092/229775445-d83ed60a-2c2d-4e6b-a82e-b35f845e20cc.png)
![image](https://user-images.githubusercontent.com/6163092/229775796-4f7c3281-cfd4-4908-967f-4a7a9fc54dc5.png)

> I don't like the new horizontal lines added. Makes the tab even more cluttered.

I was trying to group the related actions for each of these together within the groupbox:
* Save
* Load
* Delete
* Copy to other profiles.

> > How would you load the map selected in dropdown?
> 
> By selecting it from the dropdown.

By selecting it from the dropdown and THEN pressing the button to load it - you cannot just use the first action as that will fire for just scrolling through the list - and since loading the file is not trivial - nor is wiping out the one you have already in memory - it needs a positive, confirmation step before proceeding.

vadi2 commented:
re: space issue - need to think about it, will get back on this one.

> I was trying to group the related actions for each of these together within the groupbox:

I agree the the extra lines are making it look quite cluttered. It looks okay in your UI, but not in the default Linux one. 

> nor is wiping out the one you have already in memory - it needs a positive, confirmation step before proceeding.

Ah, but this is where the fact that we know if a map is saved or not comes in handy. If there were no changes to the map - why bother the user? If there were, we can pop up a dialog asking them 'you've got unsaved changes in the current map, do you want to save them first or replace the map anyway'. What do you think?

Kebap commented:
> I was trying to group

We normally use groupboxes here.
Maybe it is time to split the group "Map files" into several?
Could even reduce label texts by moving descriptions to groupbox names:
- Save map
- Load map
- Delete map

Seems like "map version" belongs to "save map" thematically? It will influence how maps are saved, not loaded or deleted.
Then again "copy to other profile" is also a way to save a map, or to load a map (in the other profile), so could go in both.

vadi2 commented:
I'm happy to consider a redesign it but that seems like it's increasing the scope of the PR, perhaps that should be done separately?

Kebap commented:
> That QComboBox needs to be wider 

I don't remember the ComboBox for profile history to be quite as wide.
So you want to keep the ultra long localized version of the day of the week like "Wednesday"..?
Surely we don't need a long version of month like "January", how about "Jan" or simply "01" to save space?
Also seconds seem a bit superfluous. Maybe they could show in dropdown when clicked upon but not on first view.
Then again, I'd rather not optimize the view to acommodate any "odd files with arbitrary names that might be put in that location"

Sorry for paining your shed. 😊

SlySven commented:
> Ah, but this is where the fact that we know if a map is saved or not comes in handy. If there were no changes to the map - why bother the user? If there were, we can pop up a dialog asking them 'you've got unsaved changes in the current map, do you want to save them first or replace the map anyway'. What do you think?

Not as a `QMessageBox` I hope! Anyhow I am not sure I understand the User interaction/experience you want. As this PR currently is:
* the user clicks on the combo box and selected a previously saved map (which could include the `autosave.dat` from a couple of minutes ago if they have made any changes) - alternatively they can also use a (vertical) mouse wheel to scroll through them without clicking!
* they then click the button and wait for the map to load - with an additional progress bar appearing over the preferences form should they chose an `.xml` or `.json` file.
* the map is available thereafter

You seem to be saying:
* the user touches the combobox in any way - including tabbing though it with the keyboard.
* if the user has any unsaved changes: a "do you want to load this map" (which by default will start out on the most recently saved map - which is probably the one loaded with the profile).
* if the user has not got any unsaved changes OR if they okay in the previous step then the map is loaded with the progress dialogues if appropriate
* if they adjust the combobox again a different map file will be selected and loaded - and so on until they get the one they wanted (or not) and smash the keyboard in frustration.

If you want to see this sort of UI action then try changing the Main display font in real time - that responds in that manner (and with a mouse wheel) - in THAT case it is appropriate even though there can be a tiny delay (at least in the debug builds with a lot of fonts) because it enables the end-user to see interactively what the choice they make looks like and because it makes a significant change to the user environment. I don't think the same is true for this...

SlySven commented:
> > That QComboBox needs to be wider
> 
> I don't remember the ComboBox for profile history to be quite as wide. So you want to keep the ultra long localized version of the day of the week like "Wednesday"..? Surely we don't need a long version of month like "January", how about "Jan" or simply "01" to save space? Also seconds seem a bit superfluous. Maybe they could show in dropdown when clicked upon but not on first view. Then again, I'd rather not optimize the view to acommodate any "odd files with arbitrary names that might be put in that location"
> 
> Sorry for paining your shed. :blush:

"I don't remember the ComboBox for profile history to be quite as wide."

--> it wasn't - and I was annoyed at the information that was hidden so I put in code to make it as wide as it needed to be to show the information for the "Connect profiles" dialogue case  - but it is NOT in this remaining case in the "Profile preferences" one.

"So you want to keep the ultra long localized version of the day of the week like "Wednesday"..? Surely we don't need a long version of month like "January", how about "Jan" or simply "01" to save space? Also seconds seem a bit superfluous."

--> That choice of long localized was already there - and there is only that or short (there is an even shorted one in some locales but in others it is the same as the short one)

"Also seconds seem a bit superfluous. Maybe they could show in dropdown when clicked upon but not on first view.
Then again, I'd rather not optimize the view to acommodate any "odd files with arbitrary names that might be put in that location"

--> Making the `QComboBox` show two different forms might be possible - but it make the code more complex and I wasn't motivated to do that. Besides changing the display format makes the UI less smooth IMHO - if I chose something it is more reassuring that what I choose stays on display rather than having it changed to a different form "just to fit in". Whilst it is not ideal to have the control to accommodate any length of text in case of "pathological" ones - I think it is the recommendation:
![image](https://user-images.githubusercontent.com/6163092/229789051-b32e3e9b-2403-40a7-801d-cc8efe75aa1c.png)

Note that this `enum` is unstable - the values are different in the above  (from Qt 5.15.2) for Qt 5.14 AND also Qt 6.4.1.

As it is, the drop down box expands to accommodate the longest entry automatically and if there is not enough space the text is truncated on the right hand side should the widget be constrained from expanding to fit when collapsed - which might be the bit of information you want to see depending on your locale's format.

vadi2 committed:
Adjust load map UI

vadi2 commented:
Alright, fair, that's not really better.

Still need to improve on the current design and I think I have a simple tweak to make it better.

# Before
![image](https://user-images.githubusercontent.com/110988/229810373-5f6faf5a-b73a-49f6-97fb-d9d9107f764f.png)

# After
![image](https://user-images.githubusercontent.com/110988/229810402-b6187675-2eab-4155-b7d1-dbef2839f184.png)

SlySven committed:
Revise: clean up changes made by Vadi to the profile_perferences.ui file

Also tweak some related texts so they are in the same format.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

SlySven commented:
How it is now:
![image](https://user-images.githubusercontent.com/6163092/230178943-5c69a785-0f72-4d8b-8828-dae68e6dda8a.png)

Note that I will need to update the `mudlet_en_US.ts` translation file manually as I put a numerus one into the "copy to" button - updated as the `pick destinations ...` control is used...

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -1907,7 +1907,7 @@ you can use it but there could be issues with aligning columns of text</string>
            <widget class="QPushButton" name="pushButton_saveMap">
             <property name="text">
-             <string>Press to choose location and save</string>
+             <string>choose location ...</string>
```
```suggestion
             <string>choose location...</string>
```

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -1934,11 +1934,42 @@ you can use it but there could be issues with aligning columns of text</string>
            <widget class="QPushButton" name="pushButton_loadMap">
             <property name="text">
-             <string>Press to choose file and load</string>
+             <string>choose map ...</string>
```
```suggestion
             <string>choose map...</string>
```

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -1991,11 +2015,11 @@ you can use it but there could be issues with aligning columns of text</string>
             </property>
             <property name="text">
-             <string>Press to pick destination(s)</string>
+             <string comment="text on button to select other profiles to receive the map from this profile, this is used when no profiles have been selected">pick destinations ...</string>
```
```suggestion
             <string comment="text on button to select other profiles to receive the map from this profile, this is used when no profiles have been selected">pick destinations...</string>
```

SlySven replied:
I've just been reading up on this - and it seems that your proposal is (one) American style - and that it varies across locales... https://en.wikipedia.org/wiki/Ellipsis#In_different_languages - of note is https://en.wikipedia.org/wiki/Ellipsis#Usage_in_computer_system_menus - which does suggest it is appropriate to use them as we have here to suggest that a further dialogue will happen on activation/clicking rather than an operation being carried out directly. :grin:

Kebap replied:
Actually we've been there before. For example only some of the actions in the mapper's context menu end in ellipsis. That is deliberate. 😉

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -2957,12 +3088,27 @@ void dlgProfilePreferences::slot_chosenProfilesChanged(QAction* _action)
     if (selectionCount) {
         pushButton_copyMap->setEnabled(true);
-        pushButton_chooseProfiles->setText(tr("%1 selected - press to change").arg(selectionCount));
+        pushButton_chooseProfiles->setText(tr("%n selected - change destinations ...",
```
```suggestion
        pushButton_chooseProfiles->setText(tr("%n selected - change destinations...",
```

vadi2 reviewed (approved):
LGTM :)

Lets give others a chance to review the final changes (2 days) and pitch in before merging.

SlySven committed:
Revise: tweak some ellipsis...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

mudlet-machine-account committed:
Merge in changes from development in a workable manner

Also, for the `./src/ui/profile_preferences.ui` file:
* fix some `QGridLayout` items so that they are back in order
* fix tab-ordering so that it is reasonable (top to bottom, column by
column, group-by-group and finally tab-by-tab)
* add some missing "buddy" links from QLabels to associated controls

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I have a separate PR to resort the elements of some `QGridLayout` layouts in the `dlgProfilePreferences.ui` file which was modified by #6737 - which prevents this PR from being merged at this time. However that PR is on a different PC that is running a different OS to the one that I am currently on so I can't fix this immediately but will try and sort it out in the next 48 hours...

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

Resolved conflict in:
* src/ui/profile_preferences.ui

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_allowHistoricalMapsToBeSelectOnOpeningProfile

SlySven commented:
Finally! I've got the PR to be merge-able... :relieved:
